tests/s3_data_plane_tests.c

```diff
@@ -2519,6 +2520,121 @@ static int s_test_s3_default_fail_headers_callback(struct aws_allocator *allocat
+        .meta_request_type = AWS_S3_META_REQUEST_TYPE_DEFAULT,
+        .headers_callback = s_s3_test_headers_callback_check,
+        .validate_type = AWS_S3_TESTER_VALIDATE_TYPE_EXPECT_FAILURE,
+        .default_type_options =
+            {
+                .mode = AWS_S3_TESTER_DEFAULT_TYPE_MODE_GET,
+            },
+        .get_options =
+            {
+                .object_path = invalid_path,
+            },
+    };
+
+    ASSERT_SUCCESS(aws_s3_tester_send_meta_request_with_options(NULL, &options, &meta_request_test_results));
+    ASSERT_TRUE(aws_atomic_load_int(&s_test_headers_callback_invoked) == 1);
+    ASSERT_TRUE(meta_request_test_results.finished_error_code == AWS_ERROR_S3_INVALID_RESPONSE_STATUS);
+
+    aws_s3_meta_request_test_results_clean_up(&meta_request_test_results);
+
+    return 0;
+}
+
+AWS_TEST_CASE(
+    test_s3_get_object_invoke_headers_callback_on_error,
+    s_test_s3_get_object_invoke_headers_callback_on_error)
+static int s_test_s3_get_object_invoke_headers_callback_on_error(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+
+    struct aws_s3_meta_request_test_results meta_request_test_results;
+    AWS_ZERO_STRUCT(meta_request_test_results);
+    aws_atomic_init_int(&s_test_headers_callback_invoked, 0);
+
+    struct aws_byte_cursor invalid_path = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("___INVALID_PATH___");
+
+    struct aws_s3_tester_meta_request_options options = {
+        .allocator = allocator,
+        .meta_request_type = AWS_S3_META_REQUEST_TYPE_GET_OBJECT,
+        .headers_callback = s_s3_test_headers_callback_check,
+        .validate_type = AWS_S3_TESTER_VALIDATE_TYPE_EXPECT_FAILURE,
+        .get_options =
+            {
+                .object_path = invalid_path,
+            },
+    };
+
+    ASSERT_SUCCESS(aws_s3_tester_send_meta_request_with_options(NULL, &options, &meta_request_test_results));
+    ASSERT_TRUE(aws_atomic_load_int(&s_test_headers_callback_invoked) == 1);
+    ASSERT_TRUE(meta_request_test_results.finished_error_code == AWS_ERROR_S3_INVALID_RESPONSE_STATUS);
+
+    aws_s3_meta_request_test_results_clean_up(&meta_request_test_results);
+
+    return 0;
+}
+
+AWS_TEST_CASE(
+    test_s3_put_object_invoke_headers_callback_on_error,
+    s_test_s3_put_object_invoke_headers_callback_on_error)
+static int s_test_s3_put_object_invoke_headers_callback_on_error(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+
+    struct aws_s3_meta_request_test_results meta_request_test_results;
+    AWS_ZERO_STRUCT(meta_request_test_results);
+    aws_atomic_init_int(&s_test_headers_callback_invoked, 0);
+
+    struct aws_s3_tester_meta_request_options options = {
+        .allocator = allocator,
+        .meta_request_type = AWS_S3_META_REQUEST_TYPE_PUT_OBJECT,
+        .headers_callback = s_s3_test_headers_callback_check,
+        .validate_type = AWS_S3_TESTER_VALIDATE_TYPE_EXPECT_FAILURE,
+        .put_options =
+            {
+                .object_size_mb = 1,
```

Comment:
The default client part size is 8 MB, when we send put with size less than part size, we will not create auto_put meta request as we don't need to split the request. Check code [here](https://github.com/awslabs/aws-c-s3/blob/main/source/s3_client.c#L812-L820). So, your test can pass as the default meta request is used here. Maybe similar issue for the get one. I will leave it to you to dig into it😬

Change the size to 10, and you solution here will fail.

Reply:
Also, as @JonathanHenson mentioned before, we don't have unit tests :(. It's hard to test the error from each type of put requests. It's a real concern... But, if you chose my suggestion, we can add tests like when the request has a success response code, and user error out the request from callback, make sure our headers callback will not be invoked again.

Reply:
> I think it's fine to invoke the headers callback for error right before the finish callback.

That's a good approach, thanks! Working on it.

Reply:
> The default client part size is 8 MB. When we send put with size less than part size, we will not create auto_put meta request as we don't need to split the request. Check code here. So, your test can pass as the default meta request is used here. 

Great catch, that explains why the Put meta request test is passing. Changed `object_size_mb` to 10 MB as suggested and the test fails. Fixing...